Python 3.8

Hoikas opened the pull request:
This updates the game engine and scripts to support Python 3. Note that the minimum required Python version is ~3.7~3.8 ~due to the `PyUnicode` API changing to properly return `const char*`~ because Python.pak support requires the new initialization API from 3.8. From my limited testing, the game plays correctly, however a full playthough was not done.

This changeset deletes the in-repo copy of the Python stdlib. We ought to not be versioning external code that changes based off the Python version. The stdlib and the rest of the scripts are now copied to the runtime output directory at build time. Hopefully this will make generating working game clients easier.

This DOES NOT:
- ~update the resource.dat generation scripts to Python 3. Per @Deledrius and my experimentation, these rely on fragile modules that are nearly impossible to trivally `pip install` on Python 3.~
- attempt to deal with the chaos of the duplicated `get[thing]W()` methods in the Python API

~Note this extends the work done in #596 because  it seems silly to replicate these changes on top of the other API changes found therein.~

Hoikas commented:
:shipit:

Hoikas commented:
I have pushed a fix for the issue in which junked python.pak files are created by plPythonPack.exe. Using python.pak is still broken however due to changes in Python 3's import machinery... Work will continue on this.

zrax commented on Scripts/Python/plasma/Plasma.py:
```diff
@@ -1468,10 +1448,6 @@ def useEAXAcceleration(self,state):
 
-    def useHardwareAcceleration(self,state):
-        """Enables or disables audio hardware acceleration."""
-        pass
```
Some of these changes don't look related to Python3-ification -- possibly from your audio PR?

Hoikas replied:
Yes, I wrote this on top of that

zrax replied:
Hmm, that does make it harder to review then...  I may hold off until that one is merged.

zrax commented on Scripts/Python/plasma/PlasmaTypes.py:
```diff
@@ -675,7 +675,7 @@ def __init__(self):
         # if we haven't got the parent keys yet, then add them to the dict
         if not self.gotParentKeys:
-            for anim in dict.values(self):
+            for anim in list(dict.values(self)):
```
`list` unnecessary for iteration

Hoikas replied:
This has been observed to occur while the dict is changing.

zrax commented on Scripts/Python/xEventTrigger.py:
```diff
@@ -85,7 +84,7 @@ def OnSDLNotify(self,VARname,SDLname,playerID,tag):
                     # are we paging things in?
                     if PageNames.value:
-                        names = string.split(PageNames.value,",")
+                        names = PageNames.values.split(",")
```
`value`

zrax commented on Sources/Plasma/FeatureLib/pfPython/cyMiscGlue4.cpp:
```diff
@@ -713,19 +693,15 @@ PYTHON_GLOBAL_METHOD_DEFINITION(PtSetBehaviorNetFlags, args, "Params: behKey, ne
-    char* nameIn = nullptr;
-    if (!PyArg_ParseTuple(args, "es|es", "utf8", &emailIn, "utf8", &nameIn))
+    ST::string email;
+    ST::string name;
```
`name` defaults to `"Friend"` in the removed code below...

Hoikas commented on CMakeLists.txt:
```diff
@@ -67,9 +67,8 @@ set_package_properties(OpenAL PROPERTIES
-set(Python_ADDITIONAL_VERSIONS 2.7)
-find_package(PythonLibs REQUIRED)
-set_package_properties(PythonLibs PROPERTIES
+find_package(Python3 REQUIRED Interpreter Development)
```
I can do that when I get the python.pak issues fixed. There is one case where I was being cute and used f-strings to fix a string issue, but that's not strictly needed.

Hoikas replied:
^^^ was intended to be a response to the overall review comment

Hoikas commented on Scripts/Python/ki/__init__.py:
```diff
@@ -2483,9 +2483,9 @@ def ChangeFontSize(self, new):
+            fontRange = list(range(len(kChat.FontSizeList) - 1))
         elif new == -1:
-            fontRange = range(len(kChat.FontSizeList) - 1, 0, -1)
+            fontRange = list(range(len(kChat.FontSizeList) - 1, 0, -1))
```
Since it's a konstant, sounds right...

Hoikas commented:
Updated to fix issues from review and add proper PEP 451 compliant import machinery for python.pak.

Hoikas commented:
Rebased onto master, which now includes two commits that were previously part of this PR. See #651.

Deledrius commented on Sources/Plasma/Apps/plClient/CMakeLists.txt:
```diff
@@ -18,15 +18,15 @@ if(MOVIE_AVAILABLE)
 )
 
-if(PYTHONINTERP_FOUND)
+if(Python2_Interpreter_FOUND)
```
This no longer matches the endif on L#37, causing a CMake Warning.

Deledrius commented on Scripts/Python/xIniAudio.py:
```diff
@@ -323,9 +323,9 @@ def SetAudioMode(init, device, eax):
+            entryDev.setValue(0, "\"" + device.encode("utf-8") + "\"")
         else:
-            gIniFile.addEntry("Audio.SetDeviceName \"" + device + "\"")
+            gIniFile.addEntry("Audio.SetDeviceName \"" + device.encode("utf-8") + "\"")
```
`TypeError: can only concatenate str (not "bytes") to str`

Hoikas commented:
9882e88  will be stripped out when #653 is merged.

Hoikas commented:
Pushed fixes for: Relto Smokers (#654), Relto Trees, and Kadish traceback on link-in (#655).

Hoikas commented:
Rebased and removed @Deledrius pinger.

zrax commented on Sources/Plasma/FeatureLib/pfPython/pyGlueHelpers.cpp:
```diff
@@ -45,31 +45,41 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 {
-    return (PyString_Check(obj) || PyUnicode_Check(obj));
+    if (PyUnicode_Check(obj)) {
+        *((ST::string*)str) = PyUnicode_AsSTString(obj);
```
I'd probably prefer `reinterpret_cast` here, but that's just a style preference...

zrax commented on Sources/Plasma/FeatureLib/pfPython/pyVaultMarkerGameNodeGlue.cpp:
```diff
@@ -130,17 +130,13 @@ PYTHON_METHOD_DEFINITION(ptVaultMarkerGameNode, setMarkers, args)
-        marker.pos = pyPoint3::ConvertFrom(pos)->fPoint;
-        marker.description = PyString_AsStringEx(desc);
-        collector.push_back(marker);
+        collector.push_back({ PyLong_AsUnsignedLong(id), PyUnicode_AsSTString(age),
+                              pyPoint3::ConvertFrom(pos)->fPoint, PyUnicode_AsSTString(desc) });
```
`emplace_back`?

Hoikas replied:
I tried that, and it resulted in a compile error :sad:

Hoikas replied:
... It helps if `VaultMarkers` has a constructor, doesn't it? Sigh.

zrax replied:
Hmm, it looks like you can do it as `vector.emplace_back(StructType { ... })`...  However, a quick test shows that that version isn't any more optimal than the `push_back` with optimization turned on :(.

It looks like this will get fixed in C++20, but for now I think it's fine to leave the `push_back`.

zrax replied:
Or just add a constructor like you did -- that works too :)

Deledrius commented on Sources/Plasma/Apps/plClient/CMakeLists.txt:
```diff
@@ -17,15 +17,7 @@ if(MOVIE_AVAILABLE)
-)
-
-if(PYTHONINTERP_FOUND)
+if(Python3_Interpreter_FOUND)
```
This no longer matches the endif on L#30, causing a CMake Warning.

Deledrius commented:
Missing is the change from `PYTHON_EXECUTABLE` to `Python3_EXECUTABLE` at https://github.com/H-uru/Plasma/blob/master/Sources/Plasma/Apps/plClient/external/CMakeLists.txt#L17-L20.  Without this, VS annoyingly launches the script using whatever the Windows association default action is (typically a text editor) rather than running the script.  Feel free to cherry-pick https://github.com/H-uru/Plasma/commit/0a8972ffc21dc003e50afaca60e880a2c7c29d96. ;)

Hoikas commented:
Fixed resource generation issues.

Hoikas commented:
Updated to fix an issue with the KI displaying player folders and editing marker quests.

Hoikas commented:
Yes, I saw those after @Deledrius pointed them out. I decided to not fix those issues in this pull request. I want this PR to be limited to the set of changes required for the game to function as intended on Python 3.7 and Python 3.8--mostly to help facilitate review.

More specifically, `PyUnicode_GetSize` seems to (mostly) be used in the "W" variants of the old, duplicated string/unicode setters in the API. Those are no longer needed, but, as mentioned above, I was hoping to tackle unifying all that and fixing the deprecation in another PR Further, I believe I have eradicated all uses script-side of `PyObject_AsReadBuffer` since `buffer()` is now dead, but again, I wanted to defer removal of those methods from the API itself to another PR.

zrax commented:
Ah, makes sense...  I didn't look closely at where they were used, I just saw them in the compiler output while checking if Python 3.7 fixed the `register` issue (spoiler alert: it does)

Hoikas commented:
Pushed a fix to _not_ generate a `__pycache__.pak` file in plPythonPack, and enabled the `Py_IsolatedFlag` in external releases, preventing the importing of modules from the `sys.path`.

Hoikas commented:
Bumped to Python 3.8. I have changed this PR to a draft until vcpkg is updated to Py 3.8 and the python.pak machinery is tested again. Rationale copy-pasta'd from the appropriate commit.

This is required due to reports from multiple testers that the client exits on loading StartUp. The issue is that Python 3 loads modules during its initialization--before we have installed our PEP 451 import machinery for python.pak. The previous code worked for us because we have Python 3 installed on our systems, which is not a reasonable assumption to have for our players. Python 3.8 adds APIs that allow us to initialize just the "core" and use the API, then initialize the interpreter itself once we have added the import machinery.

Hoikas commented:
Rerequesting review specifically for b82d8d3.

Hoikas commented:
Reopening due to the merge of python 3.8 into vcpkg and successful testing of python.pak.

Deledrius commented on README.md:
```diff
@@ -24,7 +24,7 @@ Plasma currently requires the following third-party libraries:
 - nVidia PhysX 2.6.4 - https://web.archive.org/web/20170501130625/http://developer.download.nvidia.com/PhysX/2.6.4/PhysX_2.6.4_SDK_Core.exe
 - Microsoft DirectX SDK - https://www.microsoft.com/en-us/download/details.aspx?id=6812
-- Python 2.7 - http://www.python.org/download/releases/2.7/
+- Python 3.7 (or higher) - https://www.python.org/downloads/
```
This should be bumped in 05b4deda4f1eb3da4186116d15be79d2486ffb3b.

Deledrius commented on CMakeLists.txt:
```diff
@@ -59,7 +59,7 @@ set_package_properties(OpenSSL PROPERTIES
 )
 
-set(Python_ADDITIONAL_VERSIONS 2.7)
+set(Python_ADDITIONAL_VERSIONS 3.7)
```
This should be bumped in 05b4ded.

Hoikas replied:
This diff is outdated--the Python version is no longer supplied to the CMake finder.

Deledrius replied:
Ah, good.  I didn't see it updated but also didn't see it was gone!

Deledrius commented on Sources/Plasma/FeatureLib/pfPython/cyPythonInterface.cpp:
```diff
@@ -464,16 +464,16 @@ std::vector<std::pair<std::string, std::string> > DebuggerCallback::GenerateLoca
 
-                std::string keyStr = PyString_AsString(PyObject_Str(key));
+                std::string keyStr = PyUnicode_AsUTF8(PyObject_Str(key));
                 if (keyStr == "__builtins__")
```
Should this be `"builtins"`?

Hoikas commented:
Updated to fix issues from review.

Hoikas commented:
Blocking on H-uru/PlasmaPrefix#4.

Hoikas commented:
@Zrax, @dpogue Is there anything else I need to do here?

Deledrius commented:
This looks ready to merge.  Any objections?

Hoikas commented:
:boom: :boom: :boom:
